Add the environment variables editor to Deploy Image creation

jcaianirh opened the pull request:
Add env editor to deploy-image page.

Question @spadgett, @rhamilto: How would you like to handle changes in Namespace after the editor appears. The config maps and secrets are usually loaded based on Namespace in the editor, so currently no filtering is done and all are available. Do we want the Namespace to cause an event to update the config map/secret choices and if so what happens to entries already created? Are you ok with showing all of them? Should the Namespace change cause a reset and rerender of the current editor?

When referencing the previous implementation, I noted that only env variables were available, so envFrom vars are intentionally not available. Assuming thats correct :) should the Single Values `<h3>` element be removed or should envFrom actually be included along with both sub headings?

@openshift/team-ux-review

![env-deployimage](https://user-images.githubusercontent.com/35978579/47260858-5539d500-d491-11e8-8a6c-764f2eb5d9cf.gif)

<img width="1223" alt="screen shot 2018-10-20 at 5 52 32 pm" src="https://user-images.githubusercontent.com/35978579/47260861-5d921000-d491-11e8-8b1a-3f01e736e6ac.png">

Updated screenshot 10/24/2018:
![screen shot 2018-10-24 at 5 05 49 pm](https://user-images.githubusercontent.com/35978579/47461610-2c765000-d7af-11e8-8f9a-d6d7a6f77f24.png)

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -48,9 +49,12 @@ const ImagePorts = ({ports, name}) => <React.Fragment>
   constructor (props) {
     super(props);
-
+    this.save = this.save.bind(this);
```
This isn't needed. `save` is already bound to `this` since it's declared using `=>` below.

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -359,6 +371,17 @@ export const EnvironmentPage = connect(stateToProps)(
+     * action.
+     *
+     */
+    exportEnvironment() {
```
Is it possible instead to have an optional `onChange` callback prop that returns the new values when they change? That seems a lot cleaner to me.

spadgett commented:
I might remove the `Single Values` header since it's the only option here, particularly if we go with simple name / value pairs.

jcaianirh commented:
@spadgett updated with callback suggestion, removed the Single Values heading, and removed the config map secret option. Also moved the help popup from the removed header to the Environment Variables label. Attached a new screenshot to show the updates.

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -343,48 +351,63 @@ export class DeployImage extends React.Component<DeployImageProps, DeployImageSt
+                  <div>Define environment variables as key-value pairs to store configuration settings. Drag and drop environment variables to change the order in which they are run. A variable can reference any other variables that come before it in the list, for example <code>FULLDOMAIN = $(SUBDOMAIN).example.com</code>.</div>} />
+                <div>
+                  <EnvironmentPage
+                    obj={{}}
```
We should make `obj` optional instead of passing `{}`

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -343,48 +351,63 @@ export class DeployImage extends React.Component<DeployImageProps, DeployImageSt
+                <div>
+                  <EnvironmentPage
+                    obj={{}}
+                    rawEnvData={{}}
```
I'd make this optional, too.

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -343,48 +351,63 @@ export class DeployImage extends React.Component<DeployImageProps, DeployImageSt
+                    rawEnvData={{}}
+                    envPath={['spec','template','spec','containers']}
+                    readOnly={false}
+                    dispatchEnvChanges={this.incomingEnvironmentChange} />
```
Maybe just `onChange`?

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -141,6 +144,10 @@ export class DeployImage extends React.Component<DeployImageProps, DeployImageSt
       }));
   };
 
+  incomingEnvironmentChange = (env) => {
```
onEnvironmentChange?

spadgett commented on frontend/public/components/utils/name-value-editor.jsx:
```diff
@@ -58,7 +58,9 @@ export const NameValueEditor = withDragDropContext(class NameValueEditor extends
-    const {nameString, valueString, addString, nameValuePairs, allowSorting, readOnly, nameValueId, configMaps, secrets, envFrom} = this.props;
+    const {nameString, valueString, addString, nameValuePairs, allowSorting, readOnly, nameValueId, configMaps, secrets, envFrom, isCreate} = this.props;
+    //Don't show Add for config maps/secrets unless we are allowing sorting (env editor not annotations) but not if we are on a create page or envFrom
+    const showAddConfigMapSecret = allowSorting && !isCreate && !envFrom;
```
This should be an explicit prop rather than guessing based on values like `allowSorting`.

spadgett commented:
/assign

jcaianirh commented:
@spadgett Made review edits. Also, I removed env from state in deploy-image in favor of using a class var so as not to rerender on every change.

spadgett commented on frontend/public/components/deploy-image.tsx:
```diff
@@ -63,19 +65,25 @@ export class DeployImage extends React.Component<DeployImageProps, DeployImageSt
     };
   }
 
+  env: object = {};
```
Probably worth a comment explaining why this isn't in state

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -421,14 +433,15 @@ export const EnvironmentPage = connect(stateToProps)(
             </h3>
-            <NameValueEditorComponent nameValueId={containerIndex} nameValuePairs={envVar[EnvType.ENV]} updateParentData={this.updateEnvVars} addString="Add Value" nameString="Name" readOnly={readOnly} allowSorting={true} configMaps={configMaps} secrets={secrets} />
+            }
+            <NameValueEditorComponent nameValueId={containerIndex} nameValuePairs={envVar[EnvType.ENV]} updateParentData={this.updateEnvVars} addString="Add Value" nameString="Name" readOnly={readOnly} allowSorting={true} configMaps={configMaps} secrets={secrets} addConfigMapSecret={!currentEnvVars.isCreate} />
```
> `addConfigMapSecret={!currentEnvVars.isCreate}`

I'd also make this a `addConfigMapSecret` prop passed to the `EnvironmentPage` component. I think this should be set by the parent component without making assumptions about context here.

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -298,6 +309,7 @@ export const EnvironmentPage = connect(stateToProps)(
         success: null,
         modified,
       });
+      onChange(currentEnv.dispatchNewEnvironmentVariables());
```
Do we need to check if `onChange` is defined?

jcaianirh replied:
@spadgett we don't because i defaulted onChange to a noop function from lodash

spadgett replied:
OK, I'd still suggest

```suggestion
      _.isFunction(onChange) && onChange(currentEnv.dispatchNewEnvironmentVariables());
```

Then remove the defaulting. That way, we don't run the `dispatchNewEnvironmentVariables` logic if there is no `onChange` prop

jcaianirh commented:
@spadgett completed review comment updates.

jcaianirh commented:
@spadgett squashed, thanks!

jcaianirh commented:
/retest

spadgett commented:
jenkins rebuild
